sklearn/preprocessing/imputation.py

```diff
@@ -374,3 +375,307 @@ def transform(self, X):
+        fitted_X = self._fitted_neighbors._fit_X
+        if X.shape[1] != fitted_X.shape[1]:
+            raise ValueError("Incompatible dimension between the fitted "
+                             "dataset and the one to be transformed.")
+        mask = _get_mask(X, self.missing_values)
+        n_rows_X, n_cols_X = X.shape
+        row_total_missing = mask.sum(axis=1)
+        if not np.any(row_total_missing > 0):
+            return X
+
+        # Check for excessive missingness in rows
+        bad_rows = row_total_missing > (mask.shape[1] * self.row_max_missing)
+        if np.any(bad_rows):
+            warnings.warn(
+                "There are rows with more than {0}% missing values. The "
+                "missing features in these rows are imputed with column means."
+                    .format(self.row_max_missing*100))
+            X_bad = X[bad_rows, :]
+            X = X[~bad_rows, :]
+            mask = _get_mask(X, self.missing_values)
+            row_total_missing = mask.sum(axis=1)
+        row_has_missing = row_total_missing.astype(np.bool)
+
+        if np.any(row_has_missing):
+            neighbors = self._fitted_neighbors.kneighbors(
+                X[row_has_missing, :], n_neighbors=adjusted_n_neighbors)
+
+            # Get row index, distance, and weights of donors
+            knn_distances, knn_row_index = neighbors
+            # Remove self from list of donors
+            if adjusted_n_neighbors > self.n_neighbors:
+                row_index = np.arange(X.shape[0]).reshape((X.shape[0], 1))
+                row_index = row_index[row_has_missing, :]
+                not_duplicate_index = np.where(~(row_index == knn_row_index))
+                knn_row_index = knn_row_index[not_duplicate_index].reshape(
+                    (-1, self.n_neighbors))
+                knn_distances = knn_distances[not_duplicate_index].reshape(
+                    (-1, self.n_neighbors))
+            weights = _get_weights(knn_distances, self.weights)
+
+            # Vertically split sets of k-donor indices and repeat each set by
+            # missing count in the corresponding recipient row
+            knn_row_index = np.vsplit(knn_row_index, knn_row_index.shape[0])
+            row_repeats = row_total_missing[row_total_missing != 0]
+            knn_row_index = np.repeat(
+                knn_row_index, row_repeats, axis=0).ravel()
+
+            # Get column index of donors
+            row_missing_index, col_missing_index = np.where(mask)
+            knn_col_index = np.repeat(col_missing_index, self.n_neighbors)
+
+            # Calculate kNN score and impute
+            donors = fitted_X[
+                (knn_row_index, knn_col_index)].reshape((-1, self.n_neighbors))
+            donors_mask = _get_mask(donors, self.missing_values)
+            donors = np.ma.array(
+                donors, mask=donors_mask)
+            imputed = np.ma.average(donors, axis=1, weights=weights)
```

Comment:
I don't understand how weights here is the same shape as donors, since donors has had repeat applied, but `knn_distances` and `weights` has not. Have you tested weighted prediction with varying numbers of features to be filled per sample?